src/code/z_bgcheck.c

```diff
@@ -4210,126 +4212,85 @@ u32 func_800C99FC(CollisionContext* colCtx, CollisionPoly* poly, s32 bgId) {
-    return (func_800C9A4C(colCtx, poly, bgId) & 2) ? true : false;
+s32 SurfaceType_CheckWallFlag1(CollisionContext* colCtx, CollisionPoly* poly, s32 bgId) {
+    return (SurfaceType_GetWallFlags(colCtx, poly, bgId) & WALL_FLAG_1) ? true : false;
 }
 
-/**
- * SurfaceType Is Wall Flag (1 << 2) Set
- */
-s32 func_800C9AE4(CollisionContext* colCtx, CollisionPoly* poly, s32 bgId) {
-    return (func_800C9A4C(colCtx, poly, bgId) & 4) ? true : false;
+s32 SurfaceType_CheckWallFlag2(CollisionContext* colCtx, CollisionPoly* poly, s32 bgId) {
+    return (SurfaceType_GetWallFlags(colCtx, poly, bgId) & WALL_FLAG_2) ? true : false;
 }
 
-u32 func_800C9B18(CollisionContext* colCtx, CollisionPoly* poly, s32 bgId) {
+FloorProperty SurfaceType_GetFloorProperty2(CollisionContext* colCtx, CollisionPoly* poly, s32 bgId) {
     return SurfaceType_GetData(colCtx, poly, bgId, 0) >> 26 & 0xF;
 }
 
-/**
- * SurfaceType Get Floor Property
- */
-u32 func_800C9B40(CollisionContext* colCtx, CollisionPoly* poly, s32 bgId) {
+FloorProperty SurfaceType_GetFloorProperty(CollisionContext* colCtx, CollisionPoly* poly, s32 bgId) {
     return SurfaceType_GetData(colCtx, poly, bgId, 0) >> 26 & 0xF;
 }
 
-/**
- * SurfaceType Is Floor Minus 1
- */
-u32 func_800C9B68(CollisionContext* colCtx, CollisionPoly* poly, s32 bgId) {
+u32 SurfaceType_IsSoft(CollisionContext* colCtx, CollisionPoly* poly, s32 bgId) {
     return SurfaceType_GetData(colCtx, poly, bgId, 0) >> 30 & 1;
 }
 
-/**
- * SurfaceType Is Horse Blocked
- */
 u32 SurfaceType_IsHorseBlocked(CollisionContext* colCtx, CollisionPoly* poly, s32 bgId) {
     return SurfaceType_GetData(colCtx, poly, bgId, 0) >> 31 & 1;
 }
 
-u32 func_800C9BB8(CollisionContext* colCtx, CollisionPoly* poly, s32 bgId) {
+SurfaceMaterial SurfaceType_GetMaterial(CollisionContext* colCtx, CollisionPoly* poly, s32 bgId) {
     return SurfaceType_GetData(colCtx, poly, bgId, 1) & 0xF;
 }
 
-/**
- * SurfaceType Get Poly Sfx
- */
-u16 SurfaceType_GetSfx(CollisionContext* colCtx, CollisionPoly* poly, s32 bgId) {
-    s32 id = func_800C9BB8(colCtx, poly, bgId);
+u16 SurfaceType_GetSfxOffset(CollisionContext* colCtx, CollisionPoly* poly, s32 bgId) {
+    SurfaceMaterial surfaceMaterial = SurfaceType_GetMaterial(colCtx, poly, bgId);
 
-    if ((id < 0) || (id > 14)) {
-        return NA_SE_PL_WALK_GROUND - SFX_FLAG;
+    if ((surfaceMaterial < 0) || (surfaceMaterial >= ARRAY_COUNT(sSurfaceMaterialToSfxOffset))) {
+        return SURFACE_SFX_OFFSET_DIRT;
     }
-    return sSurfaceTypeSfx[id];
+
+    return sSurfaceMaterialToSfxOffset[surfaceMaterial];
 }
 
 /**
- * SurfaceType Get ? (same indexer as Get Poly Sfx)
+ * SurfaceType Can Leave Imprints in the Material
  */
-s32 func_800C9C24(CollisionContext* colCtx, CollisionPoly* poly, s32 bgId, s32 arg3) {
-    s32 id = func_800C9BB8(colCtx, poly, bgId);
+s32 SurfaceType_IsSoftMaterial(CollisionContext* colCtx, CollisionPoly* poly, s32 bgId, s32 mask) {
```

Comment:
Not following the logic behind the name here.

First, you already have SurfaceType_IsSoft to be confused by, which stores a different "softness" property altogether.
Second, I don't think this function specifically checks "material softness", I think the caller selects a property to check via the `mask` variable and the function returns either 0 if the property isn't set, or non-zero if it has said property. It just happens that the only material property is whether a footprint can be imprinted on the ground

Reply:
So are properties defined by the mask they're used? In that case, all these SurfaceType materials would just be `SurfaceType_CheckBitsF` based on the mask they're extracted from in `surfaceData`
I think how it's used should guide how it's named, and it's used for imprints on the ground and tracks related to goron rolling? (or something else, player is hard to read).

Looking at the action bits that are set to true, it's "DIRT", "SAND", "DIRT_SHALLOW", and "SNOW" which does seem to be imprintable/soft material.

It is similar to `SurfaceType_IsSoft` so either this or the other name could be reevaluated, or the comments above the functions give enough description as to what it's checking for.

Reply:
In OoT, `IsSoft` is used for dirt paths and carpets, and its only purpose is downwards raycast onto it decrease the result Y by 1 unit

So yeaaaah pretty similar idea
I could see `IsSoft` -> `IsSoftSink` maybe?

Reply:
Yeah, I remember this discussion. I more agree now that there should be a better distinction between `SurfaceType_IsSoft` and `SurfaceType_IsSoftMaterial` although they seem to have similar purposes.

`IsSoft` -> decrease downwards raycast Y by 1 unit (dirt/carpet)
`IsSoftMaterial` -> leaves an imprint when walking (dirt/sand/dirt_shallow_snow)

Maybe something like
`IsSoft` -> `IsSoftSink`
`IsSoftMaterial` -> `IsSoftImprint`

Reply:
The issue I have with the name `IsSoftImprint` is still that the function technically does not always return whether the surface material should leave an imprint on the ground. The `mask` argument is what determines what material flag gets returned.

Reply:
The only thing `mask` is "masking" are values 0 and 1, so it's a weird implementation to begin with. Either the last bit in `mask` is 0, in which this function is useless because it always returns false, or the last bit in `mask` is 1, in which it has no effect on whether `SurfaceType_IsSoftMaterial` return true or false. So I don't think it should be relevant to the name of the function

Reply:
Could definitely use a `@note` though or have that explained in the doxygen-like arg above the function

Reply:
The implementation isn't weird, it's clearly designed to be extensible; it just happens that the extensibility is wasted since only one flag ever gets returned. 

And I think you're downplaying the bigger issue with calling it `SurfaceType_IsSoftMaterial` which is that it doesn't actually do what it says on the tin unless you pass in specific `mask` arguments. I feel that avoiding that sort of inconsistency is pretty important.

To me, the code simply works better if you treat the function as a flag getter. Doing so eliminates the "magic" involved with the `mask` argument, since you'd pass in e.g. `MATERIAL_GET_IS_SOFT' instead of say 1 or TRUE or `MAKE_FUNCTION_WORK`, and you get to avoid writing "this does x, but only if you do y" gotchas.

Reply:
So to understand the alternative, the function could be called:
`SurfaceType_HasMaterialProperty(CollisionContext* colCtx, CollisionPoly* poly, s32 bgId, s32 propertyType)`
And then you'd have:
```c
#define MATERIAL_PROPERTY_SOFT_IMPRINT (1 << 0)
#define MATERIAL_PROPERTY_1 (1 << 1) // doesn't exist, just for illustration
#define MATERIAL_PROPERTY_2 (1 << 2) // doesn't exist, just for illustration
#define MATERIAL_PROPERTY_3 (1 << 3) // doesn't exist, just for illustration
```
So then when you call the function, it'd be:

```c
if (SurfaceType_HasMaterialProperty(&play->colCtx, play, bgId, MATERIAL_PROPERTY_SOFT_IMPRINT)) {
    ....
}
```
Am I following the logic right?

Reply:
And
```c
u8 sSurfaceMaterialIsSoft[SURFACE_MATERIAL_MAX] = {
    true,  // SURFACE_MATERIAL_DIRT
    true,  // SURFACE_MATERIAL_SAND
    false, // SURFACE_MATERIAL_STONE
    true,  // SURFACE_MATERIAL_DIRT_SHALLOW
    false, // SURFACE_MATERIAL_WATER_SHALLOW
    false, // SURFACE_MATERIAL_WATER_DEEP
    false, // SURFACE_MATERIAL_TALL_GRASS
    false, // SURFACE_MATERIAL_LAVA
    false, // SURFACE_MATERIAL_GRASS
    false, // SURFACE_MATERIAL_BRIDGE
    false, // SURFACE_MATERIAL_WOOD
    false, // SURFACE_MATERIAL_DIRT_SOFT
    false, // SURFACE_MATERIAL_ICE
    false, // SURFACE_MATERIAL_CARPET
    true,  // SURFACE_MATERIAL_SNOW
};
```

would become

```c
u8 sSurfaceMaterialProperties[SURFACE_MATERIAL_MAX] = {
    MATERIAL_PROPERTY_SOFT_IMPRINT, // SURFACE_MATERIAL_DIRT
    MATERIAL_PROPERTY_SOFT_IMPRINT, // SURFACE_MATERIAL_SAND
    0,                              // SURFACE_MATERIAL_STONE
    MATERIAL_PROPERTY_SOFT_IMPRINT, // SURFACE_MATERIAL_DIRT_SHALLOW
    0,                              // SURFACE_MATERIAL_WATER_SHALLOW
    0,                              // SURFACE_MATERIAL_WATER_DEEP
    0,                              // SURFACE_MATERIAL_TALL_GRASS
    0,                              // SURFACE_MATERIAL_LAVA
    0,                              // SURFACE_MATERIAL_GRASS
    0,                              // SURFACE_MATERIAL_BRIDGE
    0,                              // SURFACE_MATERIAL_WOOD
    0,                              // SURFACE_MATERIAL_DIRT_SOFT
    0,                              // SURFACE_MATERIAL_ICE
    0,                              // SURFACE_MATERIAL_CARPET
    MATERIAL_PROPERTY_SOFT_IMPRINT, // SURFACE_MATERIAL_SNOW
};
```

Reply:
yes correct

Reply:
I see. Yeah, that's fine with me

Reply:
It makes a lot more of sense with a concrete example. Thanks